02-peer-protocol.md

```diff
@@ -357,53 +367,47 @@ the byte size of the input and output counts on the transaction to one (1).
     * [`sha256`:`txid`]
     * [`u16`:`num_witnesses`]
-    * [`num_witnesses*witness_stack`:`witness_stack`]
+    * [`num_witnesses*witness_stack`:`witnesses`]
 
 1. subtype: `witness_stack`
 2. data:
-    * [`u16`:`num_input_witness`]
-    * [`num_input_witness*witness_element`:`witness_element`]
+    * [`u16`:`num_witness_elements`]
+    * [`num_witness_elements*witness_element`:`witness_elements`]
 
 1. subtype: `witness_element`
 2. data:
     * [`u16`:`len`]
-    * [`len*byte`:`witness`]
+    * [`len*byte`:`witness_element`]
 
 #### Requirements
 
 The sending node:
-  - if it has the lowest total satoshis contributed, as defined
-    by total `tx_add_input` values:
-    - MUST transmit their `tx_signatures` before the peer
-  - MUST order the `witness_stack`s by the `serial_id` of the input they
+  - if it has the lowest total satoshis contributed, as defined by total
+    `tx_add_input` values, or both peers have contributed equal amounts
+    but it has the lowest `node_id` (sorted lexicographically):
+    - MUST transmit their `tx_signatures` first
+  - MUST order the `witnesses` by the `serial_id` of the input they
     correspond to
-  - number of `witness_stack`s MUST equal the number of inputs they added
+  - `num_witnesses`s MUST equal the number of inputs they added
 
 The receiving node:
   - MUST fail the negotiation if:
-    - if they have not sent a `tx_signatures` message and they are the
-      peer with the lowest total satoshis contributed, as defined by
-      `tx_add_input` values
     - the message contains an empty `witness_stack`
-    - the number of `witness_stack`s does not equal the number of inputs
+    - the number of `witnesses` does not equal the number of inputs
       added by the sending node
     - the `txid` does not match the txid of the transaction
-  - SHOULD apply the `witness`es to the transaction and broadcast it
+    - the witnesses are non-standard
+  - SHOULD apply the `witnesses` to the transaction and broadcast it
   - MUST reply with their `tx_signatures` if not already transmitted
 
 #### Rationale
 
-The peer with the lowest total of inputs amounts must transmit its
-`tx_signatures` first. If both peers have contributed equal input amounts,
-the peer with the first `node_id` (sorted lexicographically) sends their
-`tx_signatures` first.  This gives in a strict ordering of transmission
-for any multiparty tx collaboration.
-
-`witness` is the data for a witness element in a witness stack.
-Witness elements should *not* include their length.
+A strict ordering is used to decide which peer sends `tx_signatures` first.
+This prevents deadlocks where each peer is waiting for the other peer to
+send `tx_signatures`, and enables multiparty tx collaboration.
 
-Witness data must be sorted according to the `serial_id` of
-the corresponding input.
+`witness_element` is the data for a witness element in a witness stack.
+Witness elements should *not* be prefixed with their length.
```

Comment:
You'll have to bug @niftynei on that one :sweat_smile:, I was in favor of removing this line because I find it very confusing.

What is meant here is that the `data` byte array should not contain its own length, because that is already covered by the `len` field. @niftynei felt it was useful to mention it because by default, bitcoin witness data is length-prefixed when included in a tx.

Reply:
I rephrased that in https://github.com/niftynei/lightning-rfc/pull/6/commits/f2daadb3d975d7165f5ce0e0060bb7f17f5cb10f, hopefully it's clearer?